x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -330,22 +337,96 @@ private <T> void preAnalyze(
                         enrichPolicyResolver.resolvePolicies(
                             newClusters,
                             unresolvedPolicies,
-                            ll.map(newEnrichResolution -> action.apply(indexResolution, newEnrichResolution))
+                            new IndexResolutionWrappingListener(l, indexResolution)
                         );
                         return;
                     }
                 }
-                ll.onResponse(action.apply(indexResolution, enrichResolution));
-            }), matchFields);
-        }));
+                // whatever tuple we have here (from CCS-special handling or from the original pre-analysis), pass it on to the next step
+                l.onResponse(tuple);
+            })
+            .<Tuple<EnrichResolution, IndexResolution>>andThen((l, tuple) -> {
```

Comment:
👍
This style is more readable.
Furthermore I would extract the lambda in each stage in a separate method - it improves readability and avoids casts (inferred based on the method signature).